### ISSUE TITLE
GEP:  Gateway Routability

### ISSUE THREAD

**Issue by dprotaso** [kubernetes-sigs/gateway-api#1651]
### Discussed in https://github.com/kubernetes-sigs/gateway-api/discussions/1247

<div type='discussions-op-text'>

<sup>Originally posted by **evankanderson** June 30, 2022</sup>
Currently, the `GatewayAddress` field supports only external (LoadBalanced) `Gateway` definitions. I'd like to suggest adding a `ClusterLocal` or `ClusterIP` [AddressType](https://gateway-api.sigs.k8s.io/references/spec/#gateway.networking.k8s.io/v1alpha2.AddressType) with Extended support (like the existing `IPAddress` and `Hostname` support).

One of the early feature requests for Knative was the ability to deploy an application using Knative's HTTP routing support, but make it only available within the cluster. I want to be able to specify both the "internal" (`service.namespace.svc`) and "external" (`service.namespace.example.com`) Gateways using the same `GatewayClass` on the cluster, but ensure that the "internal" service is only reachable within the cluster. This would greatly simplify deployment for users over the instructions we have today.</div>

**Comment by robscott** [comment]
Although this was on pace to be released as experimental in v0.8.0 (thanks to lots of great work by @dprotaso), the initial API review with @thockin and @khenidak raised some important questions that I don't think we have sufficient time to resolve before the v0.8.0 release.

I think the most important points are:

1. "Public" and "Private" are hard to define
2. Those become especially hard to define when combined with the incoming Kubernetes multi-network initiative

Until we can answer these questions, we've decided to pull this GEP back to provisional with these specific questions marked as "unresolved". Unfortunately that means this will miss the v0.8.0 release.

**Comment by Rycieos** [comment]
I think any abstraction we try to make here will inherently be too limiting. If we make some field that accepts either `Public` or `Private`, inherently someone will some different behavior.

For myself, I have needed to set these fields on the `Service.Spec`:
```yaml
- type
- loadBalancerClass
- loadBalancerIP
- ipFamilyPolicy
- loadBalancerSourceRanges
- sessionAffinity
```

While a `Gateway` now has the `Infrastructure.annotations` field, not all of these settings are supported as annotations, and those that are are implementation and cloud provider specific.

My current workaround is to let the `Gateway` create a `Service`, and then clone that service with the same `selector`, but set my own values for the other fields. Then I use that `Service` instead of the one the `Gateway` creates.

What has been proposed before in the linked discussion is a a ref in the `Gateway` to a `Service`, with is used either as a template for a `Service` that is created, or that is taken ownership of.

The second option, ownership, would solve another problem (which may be better suited in a separate issue) which is how can we make a `Gateway` use an already existing Load Balancer.

That said, the problem with a reference is dependency ordering.

My proposal is to add a new field, `Gateway.Spec.ServiceSpec`, which would mirror the `Service.Spec`, and would instruct the `Gateway` how to program the `Service` it creates.

**Comment by mikemorris** [comment]
> The second option, ownership, would solve another problem (which may be better suited in a separate issue) which is how can we make a Gateway use an already existing Load Balancer.

This is a dedicated open discussion over in #1687

**Comment by tculp** [comment]
I agree with @Rycieos. In particular, I would like to be able to set loadBalancerSourceRanges on created services. Right now the lack of that capability is preventing me from switching Istio deployments to use the Gateway API

**Comment by dprotaso** [comment]
@Rycieos @tculp I believe this PR where the parameterRef in the Gateway will give you the knobs you want

https://github.com/kubernetes-sigs/gateway-api/pull/2924

Maybe that's the general solution for routability - for in cluster implementations we would recommend using that

**Comment by howardjohn** [comment]
IMO a first class field is still highly desirable. `parameterRef` is a vendor-specific escape hatch; we shouldn't settle to make every vendor implement core functionality in different ways

**Comment by tculp** [comment]
In addition, fields like loadBalancerSourceRanges started out as annotations but were promoted to spec-level fields because they were generic and useful enough to be specifically included. I think a case could be made for all spec-level fields to be supported.

Imagine if Deployments didn't support a top-level pod-spec field in the template, that would be crazy.

**Comment by dprotaso** [comment]
@robscott you mentioned cluster IP gateways in the meeting today - I'd view that as a subset of this GEP. Really it's what I'm looking for. 

Can you elaborate what you're thinking?

**Comment by NFRBEZ** [comment]
I agree with @tculp 

I have the case over GKE where i would like to restrict access to the gateway to some defined IP ranges, which work by using loadBalancingSourceRanges. 
Without ability to do it from gateway definition, i have to go on my service and to update the ranges by hand, which is a non sense

**Comment by mikemorris** [comment]
@tculp @NFRBEZ I think restricting access to a Gateway based on IP ranges is more aligned with the scope of #1141 than this issue, which is focused just on assigning an in-cluster address to a Gateway instead of an external public address.

> My proposal is to add a new field, Gateway.Spec.ServiceSpec, which would mirror the Service.Spec, and would instruct the Gateway how to program the Service it creates.

@Rycieos The difficulty with this is that Gateway is _not_ defined as a resource which creates Services in the same way Deployments create Pods. Rather, an underlying Service is a (relatively common) implementation detail of some Gateway API implementations, and consolidating the user-facing configuration onto the Gateway resource should likely be the goal, while exposing implementation details which aren't part of the spec definition is likely an anti-pattern we'd prefer to avoid.

**Comment by Rycieos** [comment]
> @Rycieos The difficulty with this is that Gateway is _not_ defined as a resource which creates Services in the same way Deployments create Pods. Rather, an underlying Service is a (relatively common) implementation detail of some Gateway API implementations, and consolidating the user-facing configuration onto the Gateway resource should likely be the goal, while exposing implementation details which aren't part of the spec definition is likely an anti-pattern we'd prefer to avoid.

That is a very good point. I still think that load balancer configuration is a common use case that needs to be supported, but having a way for implementations to not support or ignore that config would be needed.

**Comment by rikatz** [comment]
/asisgn @Miciah 
for awareness

**Comment by rikatz** [comment]
I think we should revive this, and I like the scope given here: https://github.com/kubernetes-sigs/gateway-api/issues/1651#issuecomment-1804154500

```
- type
- loadBalancerClass
- loadBalancerIP
- ipFamilyPolicy
- loadBalancerSourceRanges
- sessionAffinity
```

I would remove loadBalancerIP (given the gateway.spec.addresses) and maybe argue about some of the other fields (including sessionAffinity and loadBalancerSourceRanges), but I think at least type, class (if applicable), ipFamilyPolicy and the topology-related fields (`trafficDistribution`, `externalTrafficPolicy`, `internalTrafficPolicy`) should be considered.

I am bringing this to the next Gateway API meeting to check if we still have the desire to move with this proposal

**Comment by snorwin** [comment]
@rikatz I'd like to propose expanding the scope to include the following items:
- `allocateLoadBalancerNodePorts`
- `annotations` (adding support for service specific annotations to prevent applying load balancer-related annotations to all resources, e.g., deployment)
- `sessionAffinity` (this will be crucial for HTTP/3 and UDP routing)

**Comment by dprotaso** [comment]
FYI this GEP wasn't originally about exposing all the service properties.

If what's being proposed is going to be wildly different it might be worth creating a new issue so as to not add confusion.

**Comment by rikatz** [comment]
@dprotaso right, but the part that gets me is:

```
Currently, the GatewayAddress field supports only external (LoadBalanced) Gateway definitions. I'd like to suggest adding a ClusterLocal or ClusterIP [AddressType](https://gateway-api.sigs.k8s.io/references/spec/#gateway.networking.k8s.io/v1alpha2.AddressType) with Extended support (like the existing IPAddress and Hostname support).
```

Tho I am fine with opening a new GEP for it.

@snorwin annotations are already exposed on infrastructure.annotations, and this is usually how we pass them to services creation :)

**Comment by snorwin** [comment]
@rikatz for `.spec.infrastructure.annotations` it is required (and also enforced by conformance tests) that these annotations are propagated to all generated resources (e.g., Pods, Deployments, Services). However, I’d like to avoid cases where annotations such as `service.beta.kubernetes.io/aws-load-balancer-scheme: internet-facing` end up being applied to Pods.

**Comment by youngnick** [comment]
I think that a more narrow GEP to add a visibility/scope/routability property to IP addresses in the `status.addresses` makes sense to me. Conceivably could also include topology details like zone there as well, which would allow clients to make decisions about which address to pick.

In terms of annotations, the reason we did that is, as always, because we can't say "this behaves specially for Service and only for Service" without opening the door to many other "but why can't we have this special thing for my special resource as well".
